sigmastate/src/test/scala/sigmastate/PrettyPrintErgoTreeSpecification.scala

```diff
@@ -0,0 +1,443 @@
+        |}""".stripMargin
+  }
+
+  property("group generator"){
+    val code = "{ (x: Int) => groupGenerator }"
+    val compiledTree = compile(code)
+    PrettyPrintErgoTree.prettyPrint(compiledTree) shouldBe
+      "{ ($1: Int) => Global.groupGenerator }"
+  }
+
+  // TODO: Output should be with type param, i.e. `substConstants[Any](...)`? Related to printer config (show explicit types)
+  property("substConstants"){
+    val code = "{ (x: (Coll[Byte], Int)) => substConstants[Any](x._1, Coll[Int](x._2), Coll[Any](sigmaProp(false), sigmaProp(true))) }"
+    val compiledTree = compile(code)
+    PrettyPrintErgoTree.prettyPrint(compiledTree, 84) shouldBe
+      """{ ($1: (Coll[Byte], Int)) =>
+        |  substConstants(
+        |    $1._1, Coll[Int]($1._2), Coll[SigmaProp](sigmaProp(false), sigmaProp(true))
+        |  )
+        |}""".stripMargin
+  }
+
+  property("constant placeholder"){
+    import Values.CollectionConstant
+    import Values.ByteArrayConstant
+    import scorex.util.encode.Base16
+    def stringToByteConstant(in: String): CollectionConstant[SByte.type] = ByteArrayConstant(in.getBytes("UTF-8"))
+    def decodeString(in: String): CollectionConstant[SByte.type] = ByteArrayConstant(Base16.decode(in).get)
+
+    val expectedResult = decodeString("ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad")
+    val calcSha256 = EQ(CalcSha256(stringToByteConstant("abc")), expectedResult)
+    val ergoTree = mkTestErgoTree(calcSha256.toSigmaProp)
+    // TODO: prettyPrint have to be updated as constant placeholders are not part of ErgoTree root
+  }
+
+  property("ergolike nodes"){
+    val code = 
+      """{ (x: Context) =>
+        |  val a = x.OUTPUTS.exists({ (box: Box) => box.value + 5 > 10 })
+        |  val b = x.INPUTS.map { (b: Box) => b.value }.getOrElse(5, 0L)
+        |  val c = x.HEIGHT
+        |  val d = x.LastBlockUtxoRootHash
+        |  val e = CONTEXT.dataInputs(0).tokens(0)
+        |  (a, (b, (c, (d, e))))
+        |}""".stripMargin
+    val compiledTree = compile(code)
+    PrettyPrintErgoTree.prettyPrint(compiledTree) shouldBe
+      """{ ($1: Context) =>
+        |  (
+        |    OUTPUTS.exists({ ($3: Box) => ($3.value + 5.toLong) > (10.toLong) }), (
+        |      INPUTS.map({ ($3: Box) => $3.value }).getOrElse(5.toInt, 0.toLong), (
+        |        HEIGHT, (
+        |          $1.LastBlockUtxoRootHash, CONTEXT.dataInputs(0.toInt).tokens(0.toInt)
+        |        )
+        |      )
+        |    )
+        |  )
+        |}""".stripMargin
+  }
+
+  property("function application term"){
+    val code = 
+      """{(x: Option[Long]) =>
+        |  def f(opt: Long): Long = opt + 3
+        |  if (x.isDefined) f(x.get) else f(5L)
+        |}""".stripMargin
+    val compiledTree = compile(code)
+    PrettyPrintErgoTree.prettyPrint(compiledTree) shouldBe
+      """val $1 = { ($1: Long) => $1 + 3.toLong }; { ($2: Option[Long]) =>
```

Comment:
- $1 => func1
- new line after semicolon
- top level block should be enclosed to `{` ... `}`